### DISCUSSION TITLE
Trying to proxy extended properties across relationship

### DISCUSSION THREAD

**Discussion by dcode** [sqlalchemy/sqlalchemy#9821] [Usage Questions]
Hi, I'm new to SQLAlchemy but long user of Python.  I'm trying to implement an API for an object that has some core fields, but allows arbitrary key-values as well. I've implemented the `Image` model with core properties and the `ImageProperty` model for the extended key-value pairs. I'm doing this with `flask_sqlalchemy`, so `db.Model` is a declarative base model. I've also been able to override `__getattr__` to transparently return individual properties.

```
class Image(db.Model):
    """Class for core Image properties."""
    __tablename__ = "images"
    id = db.Column(db.String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    name = db.Column(db.String(255))
    status = db.Column(db.String(30), nullable=False)
    
    properties = db.relationship("ImageProperty")
    
    def __getattr__(self, name):
      logger.info(f"Looking up attribute {name}")
      for entry in self.properties:
          if entry.name == name:
              return entry.value

      raise ValueError

class ImageProperty(db.Model):
    """Represents arbitrary key-value properties for an image in the datastore."""
    __tablename__ = "image_properties"
    __table_args__ = (
        db.UniqueConstraint("image_id", "name", name="ix_image_properties_" "image_id_name"),
    )

    id = db.Column(db.Integer, primary_key=True)
    image_id = db.Column(db.String(36), db.ForeignKey("images.id"), nullable=False)
    image = db.relationship(Image, back_populates="properties")

    name = db.Column(db.String(255), nullable=False)
    value = db.Column(db.Text)
```

This works great for `Image.query.all()`, as I have a Marshmallow schema that transparently merges into a single JSON object on output and splits on input. The problem is I want to translate the user query against the flat JSON object. Say I have a key called `os_distro`. I want to retrieve all Images that have the property of `os_distro == "fedora"`.

Nevermind, scratch the `__getattr__` thing, it breaks almost everything else.

So, is the best approach here something like check to see if `Image` has the attribute and if not,

```
images = Image.query.join(ImageProperty).filter(ImageProperty.name == query_key).filter(ImageProperty.value == query_value)
```

But if it does, how can I dynamically get the class attribute?

All of this feels like I'm over complicating it.

**Comment by CaselIT** [comment]
Hi,

Note that you can use dict as the collection of `properties` avoiding the loop to access property by name. See https://docs.sqlalchemy.org/en/20/orm/collection_api.html#dictionary-collections

That said you are probably looking for relationship.any(): https://docs.sqlalchemy.org/en/20/orm/internals.html#sqlalchemy.orm.RelationshipProperty.Comparator.any
